pkg/experiment/query_backend/query_plan/query_plan.go

```diff
@@ -77,270 +22,74 @@ func Open(p *queryv1.QueryPlan) *QueryPlan {
-			// that have a parent.
-			n := nodes[len(nodes)-1]
-			o := n.off + n.len
-			l := uint32(len(nodes)) - o
-			nodes = append(nodes, node{
-				typ: NodeMerge,
-				off: o,
-				len: l,
-			})
-			break
-		}
-		off += length
-	}
-	return &QueryPlan{
-		blocks: blocks,
-		nodes:  nodes,
+		return new(queryv1.QueryPlan)
 	}
-}
 
-// Root returns the root node of the query plan.
-func (p *QueryPlan) Root() *Node {
-	if len(p.nodes) == 0 {
-		return &Node{Type: NodeRead, p: p}
+	if maxReads < 1 {
+		return new(queryv1.QueryPlan)
 	}
-	n := Node{p: p}
-	n.n = p.nodes[len(p.nodes)-1]
-	n.Type = n.n.typ
-	return &n
-}
 
-// Plan returns the query plan scoped to the node.
-// The plan references the parent plan blocks.
-func (n *Node) Plan() *QueryPlan {
-	// BFS traversal. Our goal is to preserve the order of
-	// nodes as in the original plan, and only shift offsets.
-	nodes := make([]node, 0, 32)
-	stack := make([]node, 0, 32)
-	stack = append(stack, n.n)
-	var x node
-	for len(stack) > 0 {
-		x, stack = stack[0], stack[1:]
-		if x.typ == NodeMerge {
-			// Add child nodes in the reverse order, to honour
-			// the order of the original plan, compensating the
-			// stack LIFO at this level. We do it after append
-			// to stack to not modify the original plan.
-			s := len(stack)
-			stack = append(stack, n.p.nodes[x.off:x.off+x.len]...)
-			slices.Reverse(stack[s:])
-		}
-		nodes = append(nodes, x)
-	}
-	if len(nodes) == 0 {
-		return new(QueryPlan)
-	}
-	// Swap merge and read nodes as their order is changed
-	// during the traversal. The order of nodes within the
-	// same type is revered as well (order within the level
-	// is fixed at traversal).
-	var p NodeType // Previous node type.
-	var s int
-	for i, c := range nodes {
-		if p != 0 && c.typ != p {
-			s = i
-			break
-		}
-		p = c.typ
-	}
-	slices.Reverse(nodes[:s]) // Merge nodes.
-	slices.Reverse(nodes[s:]) // Read nodes.
-	tmp := make([]node, s)
-	copy(tmp, nodes[:s])
-	copy(nodes, nodes[s:])
-	copy(nodes[len(nodes)-s:], tmp)
-	if nodes[0].typ != NodeRead {
-		panic("bug: first node must be a read node")
+	if maxMerges < 2 {
+		return new(queryv1.QueryPlan)
 	}
 
-	// Offset correction by node type: as the plan is a subtree,
-	// the offsets of the child nodes are shifted.
-	o := [3]int{0, -1, -1} // Table of offsets by type.
-	bs := nodes[0].off     // Offset of the first referenced block.
-	for i, c := range nodes {
-		// Correct the offset.
-		if o[c.typ] < 0 {
-			// This is the first node of the type: we remember
-			// the offset and reset it to zero, as children of
-			// the first node are _always_ placed at the very
-			// beginning (for both read and merge nodes).
-			nodes[i].off = 0
-			o[c.typ] = int(c.len)
-		} else {
-			nodes[i].off = uint32(o[c.typ])
-			o[c.typ] += int(c.len)
+	// create leaf nodes and spread the blocks in a uniform way
+	var leafNodes []*queryv1.QueryNode
+	for i := 0; i < len(blocks); i += maxReads {
```

Comment:
This is an optimisation we can make later:

The breakdown is not entirely uniform, as is also the case in the source version. If we have `maxReads` = 20, and 101 blocks, the last node will have just 1 block.

I had this snippet (but I'm sure there's a better way):
```go
func Split[S ~[]E, E any](slice S, batchSize int) []S {
	size := len(slice)
	if size == 0 || batchSize < 1 {
		return nil
	}
	n := int(math.Ceil(float64(size) / float64(batchSize)))
	base := size / n
	remainder := size % n
	batches := make([]S, 0, n)
	var offset int
	for i := 0; i < n; i++ {
		end := offset + base
		if i < remainder {
			// Distribute remaining elements among the first few batches.
			end++
		}
		batches = append(batches, slice[offset:end])
		offset = end
	}
	return batches
}
```

[Example](https://go.dev/play/p/msFfGzOg5K6):
```
data   []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
batch  3

Batch 1: [1 2 3]
Batch 2: [4 5 6]
Batch 3: [7 8]
Batch 4: [9 10]
```

Probably the same for merge nodes, as far as I understand.